crypto/bls/blst/signature.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2023 The klaytn Authors
+// This file is part of the klaytn library.
+//
+// The klaytn library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package blst
+
+import (
+	"crypto/rand"
+	"fmt"
+	"sync"
+
+	"github.com/klaytn/klaytn/crypto/bls/types"
+	"github.com/pkg/errors"
+)
+
+type signature struct {
+	// Pointer to underlying blst struct, hence the name 'p'
+	p *blstSignature
+}
+
+func SignatureFromBytes(b []byte) (types.Signature, error) {
+	if len(b) != types.SignatureLength {
+		return nil, types.ErrSignatureLength(len(b))
+	}
+
+	if s, ok := signatureCache.Get(cacheKey(b)); ok {
+		return s.(*signature).Copy(), nil
+	}
+
+	p := new(blstSignature).Uncompress(b)
+	if p == nil || !p.SigValidate(false) {
+		return nil, types.ErrSignatureUnmarshal
+	}
+
+	s := &signature{p: p}
+	signatureCache.Add(cacheKey(b), s.Copy())
+	return s, nil
+}
+
+func MultipleSignaturesFromBytes(bs [][]byte) ([]types.Signature, error) {
+	if len(bs) == 0 {
+		return nil, nil
+	}
+	for _, b := range bs {
+		if len(b) != types.SignatureLength {
+			return nil, types.ErrSignatureLength(len(b))
+		}
+	}
+
+	// Separate cached and uncached element
+	sigs := make([]types.Signature, len(bs))
+	var batchIndices []int
+	var batchBytes [][]byte
+	for i, b := range bs {
+		if sig, ok := signatureCache.Get(cacheKey(b)); ok {
+			sigs[i] = sig.(*signature).Copy()
+		} else {
+			batchIndices = append(batchIndices, i)
+			batchBytes = append(batchBytes, b)
+		}
+	}
+
+	// Compute on uncached elements
+	batchPs := new(blstSignature).BatchUncompress(batchBytes)
+	if batchPs == nil || len(batchPs) != len(batchBytes) {
+		return nil, types.ErrSignatureUnmarshal
+	}
+
+	// Merge cached and uncached elements
+	for i, outIdx := range batchIndices {
```

Comment:
does outIdx start after the cached elements? i think sigs[] already has cached elements, i am just curious.

Reply:
Did you ask if outIdx avoids cached elements? then yes. Please refer to the [range syntax](https://gobyexample.com/range).

Suppose [A,B,C,D] were given and A,C were previously cached. Then
batchBytes = [B,D]
batchIndices = [1,3]
In the for loop,
- i = 0, outIdx = 1
- i = 1, outIdx = 3

Reply:
That's right. Thanks